src/Language/PureScript/TypeChecker/Roles.hs

```diff
@@ -71,47 +88,88 @@ lookupEnvRoles env tyName =
   :: forall m
-   . (MonadError MultipleErrors m, MonadState CheckState m)
-  => ModuleName
+   . (MonadError MultipleErrors m)
+  => Environment
+  -> ModuleName
   -> ProperName 'TypeName
     -- ^ The name of the data type whose roles we are checking
   -> [(Text, Maybe SourceType)]
     -- ^ type parameters for the data type whose roles we are checking
   -> [DataConstructorDeclaration]
     -- ^ constructors of the data type whose roles we are checking
   -> m [Role]
-checkRoles moduleName tyName tyArgs ctors = do
-  let qualName = Qualified (Just moduleName) tyName
-  roleMaps <- traverse (walk mempty . snd) $ ctors >>= dataCtorFields
-  let
-    ctorRoles = getRoleMap $ mconcat roleMaps
-    inferredRoles = map (\(arg, _) -> fromMaybe Phantom (M.lookup arg ctorRoles)) tyArgs
-  env <- getEnv
-  rethrow (addHint (ErrorInRoleDeclaration tyName)) $ do
-    case M.lookup qualName (roleDeclarations env) of
-      Just declaredRoles -> do
-        let
-          k (var, _) inf dec =
-            if inf < dec
-              then throwError . errorMessage $ RoleMismatch var inf dec
-              else pure dec
-        sequence $ zipWith3 k tyArgs inferredRoles declaredRoles
-      Nothing ->
-        pure inferredRoles
+checkRoles env moduleName tyName tyArgs ctors =
+  checkDataBindingGroupRoles env moduleName [(tyName, tyArgs, ctors)] tyName tyArgs
+
+type DataDeclaration =
+  ( ProperName 'TypeName
+  , [(Text, Maybe SourceType)]
+  , [DataConstructorDeclaration]
+  )
+
+checkDataBindingGroupRoles
+  :: forall m
+   . (MonadError MultipleErrors m)
+  => Environment
+  -> ModuleName
+  -> [DataDeclaration]
+  -> ProperName 'TypeName
+  -> [(Text, Maybe SourceType)]
+  -> m [Role]
+checkDataBindingGroupRoles env moduleName group =
+  let initialRoleEnv = M.union (roleDeclarations env) (getRoleEnv env)
+      inferredRoleEnv = inferDataBindingGroupRoles moduleName group initialRoleEnv
+  in \tyName tyArgs -> do
+    let qualTyName = Qualified (Just moduleName) tyName
+        inferredRoles = M.lookup qualTyName inferredRoleEnv
+    rethrow (addHint (ErrorInRoleDeclaration tyName)) $ do
+      case M.lookup qualTyName (roleDeclarations env) of
+        Just declaredRoles -> do
+          let
+            k (var, _) inf dec =
+              if inf < dec
+                then throwError . errorMessage $ RoleMismatch var inf dec
+                else pure dec
+          sequence $ zipWith3 k tyArgs (fromMaybe (repeat Phantom) inferredRoles) declaredRoles
+        Nothing ->
+          pure $ fromMaybe (Phantom <$ tyArgs) inferredRoles
+
+inferDataBindingGroupRoles
+  :: ModuleName
+  -> [DataDeclaration]
+  -> RoleEnv
+  -> RoleEnv
+inferDataBindingGroupRoles moduleName group roleEnv =
+  let (Any update, roleEnv') = flip runState roleEnv $
+        mconcat <$> traverse (state . inferDataDeclarationRoles moduleName) group
+  in if update
+     then inferDataBindingGroupRoles moduleName group roleEnv'
+     else roleEnv'
+
+inferDataDeclarationRoles
```

Comment:
Could you add some documentation about the meaning of the `Any` in the return type please?

Reply:
I added a comment and aliased the `Any` to `DidRolesChange` to make things explicit.